### ISSUE TITLE
Let googlebot check the hsts preload list to make Port 80 deprecation safer

### ISSUE THREAD

**Issue by TerraX-net** [chromium/hstspreload.org#67]
#47 "Allow servers that block port 80 to preload"
Yesterday we finally closed port 80 and the websites ever had a 301 from http to https. That's optimal.

But
1. Imagine there are many links on the web to http://old-popular-site
2. old-popular-site switches to https and uses HSTS + Preloading and closes Port 80
3. googlebot would detect the new https:// site, but could not follow the links on other websites to http://old-popular-site as ranking signal, where the browsers would directly open the preloaded https://old-popular-site
4. googlebot should know the preloading list for the case it can't follow an http-link, so it will try https like the browsers would do.

Maybe this suggestion is already implemented (?)
If you have any cons, please let me know.

**Comment by lgarron** [comment]
Is this a feature request for the Google crawler itself?

If a site was served over HTTP for a significant period of time, sending a 301 is probably best.
If a site never answered over port 80, or has made 301 canonical, search engines should be able to adapt in a matter of days to months.

I can ask whether the team in charge of the Google crawler would consider using the preload list, but I'm guessing it wouldn't be simple, and there would be cases with unexpected or undesirable behaviour.

**Comment by TerraX-net** [comment]
> I can ask whether the team in charge of the Google crawler would consider using the preload list, but I'm guessing it wouldn't be simple, and there would be cases with unexpected or undesirable behaviour.

Yes, please!
If Wikipedia still had an http-link `<a href="http://old-popular-site/folder/important-publication.html">The source of an important quote</a>`, while old-popular-site got preloaded (and closed port 80), any modern browser would follow that link, but directly to the preloaded http**S**://old-popular-site/folder/important-publication.html where googlebot might think that website is down and this link would lost it's "ranking weight"(?)

Could good content be less worth because ancient browsers does not behave like modern browsers? IE8 users probably would google "old-popular-site important publication" then and google would present the https-url.

I would be unable to suggest people to get preloaded and to close port 80 in combination (in the case they only expect modern/updated browsers anyway) if googlebot would think that resource is offline.
The whole URL (with the now to https modified protocol) would lost it's previous rank.

If someone wants to get preloaded, he has to do a 301 in every case from http to https, so he's saying "please only visit my website via https". No Browser would connect to a http://mypage/secret/file.html because of it's internal preloading list. So Googlebot should behave like a modern browser and replace http with https when it comes across an URL of a preloaded domain. (HSTS preloading = http killswitch)

**Comment by lgarron** [comment]
Are you aware of [canonical URLs](https://support.google.com/webmasters/answer/139066?hl=en)?

If you properly update the canonical URLs for a site to HTTPS before redirecting and preloading, there should be no significant ranking hit: https://plus.google.com/+JohnMueller/posts/PY1xCWbeDVC

I'm not sure if you're asking for anything that couldn't be addressed by a proper migration.

**Comment by TerraX-net** [comment]
https://plus.google.com/+JohnMueller/posts/PY1xCWbeDVC
> Do I lose "link juice" from the redirects? No, for 301 or 302 redirects from HTTP to HTTPS no PageRank is lost.

But what, if we expect that the HSTS preloading list is like a client-side 301 redirect, so there is no need to have an open port 80 on the server side.

There should be a way to run a website port 443-only after it got into the preloading list, without having an open port 80 with a 301 as long there are any links out in the world linking to the old http-url.

As I understand, "canonical" would make sense if there would be duplicate content.
But if (?) "canonical" is really that powerful as you are suggesting, then the following should be possible to address this github issue:

1. http://old-popular-site/folder/important-publication.html is no longer available, because port 80 is closed after got added to the preloading list, but there are many links to it on the web. Browsers would directly access this via https because of the preloading.
2. httpS://old-popular-site/folder/important-publication.html has <link canonical with it's https-url in it and a nice sitemap.
3. "https://old-popular-site" into google.com/addurl
3. googlebot finds those (dead) http-urls on the web (and ignores that they technically are no more accessible because port 80 is closed, but practically accessible via preloaded https) and has those "canonical"-flagged httpS-urls seen before (through google.com/addurl) and realises, that they are the same and have "weight", so there is no lost of ranking?

Is this possible and what you had in mind? We want to promote the port 80 closing process correctly without killing someones internet life.

**Comment by lgarron** [comment]
Sites should not close port 80 without preparation. To quote from the link in my previous comment:

> **What's the next step after testing?**
> Follow our site-move documentation ( https://support.google.com/webmasters/answer/6033049 ). Use 301 redirects from HTTP to HTTPS, confirm the new version by adding a rel=canonical on the HTTPS page, pointing to itself, and submit sitemaps including both HTTP & HTTPS URLs with new change-dates (in the long run, just keep the HTTPS sitemap). 

After all the HTTPS pages are indexed and returned in relevant search results, it should be safe to turn down port 80.

What is your motivation for getting sites to close port 80?
You say it's "optimal", but there is little impact on security if you get HSTS right, and most sites that are migrating to HTTPS will probably want to support non-HSTS clients to reach the most users possible.

**Comment by TerraX-net** [comment]
> Use 301 redirects from HTTP to HTTPS, confirm the new version by adding a rel=canonical 

It would be perfect, if Google's HSTS preloading list is for every modern client the client-side http-to-https-301, even for googlebot.

> After all the HTTPS pages are indexed and returned in relevant search results, it should be safe to turn down port 80.

Currently, this point is maybe 1-2 weeks after a tidy https website + 301 from http to https.
Ideally it should be enough to have rel=canonical + https sitemap + HSTS preloading.

It would be the fault of other less important bots/search bots if they don't behave like a modern webbrowser. (I still hope Googlebot will behave like a modern webbrowser and knew that client-side 301 forced by HSTS preloading).

> What is your motivation for getting sites to close port 80?

It's an idealistic¹ approach to only have ports open for encrypted communication. Maybe something like https://tools.ietf.org/html/rfc7858 will even succeed unencrypted DNS on port 53 to make the use of public WiFi hotspots more private at some time. Not everyone want's to use OpenVPN. TLS should be fine. QUIC sounds interesting.

> probably want to support non-HSTS clients to reach the most users possible.

Such clients are not modern browsers with the latest security patches. There should be a large movement of websites owners to let non-HSTS & non-TLSv1.2 clients die
and a movement on the browser vendor side to [fallback to https if port 80 is closed](https://bugzilla.mozilla.org/show_bug.cgi?id=1002724).
We have some sites with TLS 1.0 + PFS as minimum and nearly all users come via search engines (the most) or direct links (modern browsers). Users below this could google for content they want to find or could use a modern webbrowser (or boot their PC at home / buy a new tablet). They have to feel their technical weakness / vulnerability to move away from their ancient browser.

> I can ask whether the team in charge of the Google crawler would consider using the preload list

Please.

¹) That's our legal structure (non-profit idealistic association). That's why we try to find the most secure and usable configuration.

**Comment by lgarron** [comment]
From what I can tell, you are saying three things:

1. Sites should be able to operate without port 80.
2. Sites should be able to migrate to HTTPS and close port 80 instantly.
3. Search engine crawlers should behave like modern browsers, at least when it comes to HSTS and preloading.

I totally agree with 1, but I think it's reasonable if 2 takes a couple of weeks for old sites that care about their search rankings.

Point 3 is probably not simple, and I still don't see a strong reason to try to advocate for it.
If I'm going to take this request to the relevant team, I should either have a reason it should be valuable to *them*, or why it's important for the internet at large. To me, I don't see a strong argument for either.

**Comment by TerraX-net** [comment]
My thoughts so far:
* https-only webhoster could oblige existing and new customers via general terms and conditions to accept a closed port 80, so they hadn't to have special migration servers with an open port 80. -> avoid unnecessary efforts.

* Somebody is switching to https-only because he doesn't want manipulated connections to this visitors without any authenticity. keywords: "cloudflare interception india". a government could make a critical website unattractive to search bots in the transition time.

* It preserves against mis-configuration in the transitional period (google itself only has a temporary redirect from http://mail.google.com to https://mail.google.com. others could make similar mistakes on a subdomain.)

* Site operators probably would not dare to close port 80 because they could not be 100% sure after months of 301 redirection to not lose their ranks. It would be too risky for them.

We are still in the process to collect more arguments for them. I have asked Ivan Ristic to think about it and maybe he would add some points too if he also came to the conclusion that this would be useful.

**Comment by ScottHelme** [comment]
I have a blog coming up on this topic soon, but I really don't understand the reason to close port 80. Can you explain?

Whilst browsers still default to http/80 I don't see any benefit at all. Even when they start shifting, we'd still need to leave 80 open for all kinds of other clients, even if only to redirect to https/443.

**Comment by TerraX-net** [comment]
> Can you explain?
> need to leave 80 open for all kinds of other clients

https://en.wikipedia.org/wiki/KISS_principle Not open avoidable ports.

Analogy: Why should a mailserver should accept non-STARTTLS connections on port 25 at some point anymore? This will come, too! Should we wait 5 more years? Oh please not. Big mail providers (or many at the same time) should deny non-starttls connections from large game publishers etc.

An ancient browser has to feel like TLS 1.2-only in IE 8 on Windows XP: Connection error.
The question is: Why should I even open port 80? **To irresponsibly make life with security flaws in old browsers more comfortably?** Or for nasty bots which don't even follow a 301 on their search for a wordpress security flaw?
No, this issue is about speeding up (don't wait weeks) and ensuring (don't risk your ranking) a proper migration. We have the HSTS preloading list as a tool until the browsers fallback to https or check https first to overcome their current habit to check http first. Good search engines even observe the "safebrowsing" blacklist.

**Comment by ScottHelme** [comment]
I'm still not following though. 

If the older browser makes a request on port 80 they're already exposed to the risk, whether or not we have the port open is irrelevant. At least in those instances where they do hit port 80 we can redirect them to 443 and secure them. 

Why does a bot not following a 301 matter? You shouldn't be responding with anything other than a 301 on port 80 so why does it matter if they don't follow? Nothing should happen as a result of a request to port 80. 

I still can't think of a single instance where leaving port 80 open results in better security, only worse.

**Comment by TerraX-net** [comment]
> we can redirect them to 443 and secure them.

That's a lie because only redirecting them to their browser update download page would "secure" them by informing them about the big problem they have.
If users of ancient browsers would get "Connection error" more often on http-links they would realize that something is not correct with their [outdated and insecure] browser. Not the latest browser = maybe no HSTS at all (= pseudo security because the next time this browser propably won't be able to follow a 301 to https because someone is faking the response).

> Why does a bot not following a 301 matter?

Because only link preview bots (telegram), googlebot (unnecessary because it also crawled https) and bad bots (undesired) ever connected to our port 80. (with a big site of one of our members behind it: perfektesgewicht.de (translation: perfect weight)).

Yes, we had a global 301 to https inside "server_name _;" (:80,nginx).
100% of the ~5% still using TLS 1.0 (only very old Androids) are coming via Google. Desktop users usually have the latest {stable, ESR} version. Only ESR users would be affected of port 80 deprecation if they would open a http-link to a recently to https migrated website. That's a fair deal.

**Comment by ScottHelme** [comment]
It's not really a lie, is it. Getting a user to update their browser is a little out of scope for a discussion on whether or not to close port 80. The extension of your argument is they should also update their OS, apply updates, install AV etc etc... We aren't responsible for that, but what we can do is redirect them to https when they communicate via http. It's clear we don't agree so I will leave it at that :)

**Comment by TerraX-net** [comment]
>  Getting a user to update their browser is a little out of scope for a discussion on whether or not to close port 80. 

You twist my words. If a car has fundamental technical defects, it may lose it's permission to drive on public streets ( https://en.wikipedia.org/wiki/MOT_test ).
There is no intention of the state to get you buying a new car or repairing it. It's completely your own decision. If a user would have more and more problems with http links he would propably switch the device (old tablet/phone/browser at work -> computer at home).

It's easy for you to ignore the arguments itself:
>  pseudo security because the next time this browser propably won't be able to follow a 301 to https because someone is faking the response).

or 

> To irresponsibly make life with security flaws in old browsers more comfortably?

I don't get what your aim is? Is it completely different?
You: "compability with _best possible_ security, but best comfort for everyone. No losses in sales thinkable." 
We: "security with modern browsers, but broad movement to let the 5% with ancient browsers know their vulnerability"

You are trying to direct this comletely to off-topic, about the question opening or not opening port 80 itself, although this issue is about how to speed up migration from http to https.

Fulfilling this suggestion here to let googlebot know the preloading list would even help websites who leave their port 80 open. 

I don't want to repeat myself.

**Comment by TerraX-net** [comment]
A potential alternative:
It seems like Mozilla is going to implement [Fall back to HTTPS in URL fixup (including location bar) (like for hosts that refuse a connection on port 80 via HTTP, but accepts connections via HTTPS on port 443)](https://bugzilla.mozilla.org/show_bug.cgi?id=1002724), maybe this would be better here.
Each crawler could independently check https:// after a connection error on http:// when revisiting an url
(and if useful here, replace the old url in your database, if Strict-Transport-Security is greater than 0 when there is no rel=canonical).
